source/common/tcp_proxy/upstream.cc

```diff
@@ -244,31 +246,55 @@ void Http2Upstream::setRequestEncoder(Http::RequestEncoder& request_encoder, boo
   request_encoder_->getStream().addCallbacks(*this);
   const std::string& scheme =
       is_ssl ? Http::Headers::get().SchemeValues.Https : Http::Headers::get().SchemeValues.Http;
-  auto headers = Http::createHeaderMap<Http::RequestHeaderMapImpl>(
-      {{Http::Headers::get().Method, "CONNECT"},
-       {Http::Headers::get().Protocol, Http::Headers::get().ProtocolValues.Bytestream},
-       {Http::Headers::get().Scheme, scheme},
-       {Http::Headers::get().Path, "/"},
-       {Http::Headers::get().Host, hostname_}});
+  auto headers = Http::createHeaderMap<Http::RequestHeaderMapImpl>({
+      {Http::Headers::get().Method, config_.use_post() ? "POST" : "CONNECT"},
+      {Http::Headers::get().Host, config_.hostname()},
+      {Http::Headers::get().Path, "/"},
+      {Http::Headers::get().Scheme, scheme},
+  });
+
+  if (!config_.use_post()) {
+    headers->addReference(Http::Headers::get().Protocol,
+                          Http::Headers::get().ProtocolValues.Bytestream);
+  }
+
+  for (const auto& header : config_.headers()) {
+    if (header.append().value()) {
+      headers->addCopy(Http::LowerCaseString(header.header().key()), header.header().value());
+    } else {
+      headers->setCopy(Http::LowerCaseString(header.header().key()), header.header().value());
+    }
+  }
+
   const auto status = request_encoder_->encodeHeaders(*headers, false);
   // Encoding can only fail on missing required request headers.
   ASSERT(status.ok());
 }
 
 Http1Upstream::Http1Upstream(Tcp::ConnectionPool::UpstreamCallbacks& callbacks,
-                             const std::string& hostname)
-    : HttpUpstream(callbacks, hostname) {}
+                             const TunnelingConfig& config)
+    : HttpUpstream(callbacks, config) {}
 
 void Http1Upstream::setRequestEncoder(Http::RequestEncoder& request_encoder, bool) {
   request_encoder_ = &request_encoder;
   request_encoder_->getStream().addCallbacks(*this);
 
   ASSERT(request_encoder_->http1StreamEncoderOptions() != absl::nullopt);
+  ASSERT(!config_.use_post());
+
   auto headers = Http::createHeaderMap<Http::RequestHeaderMapImpl>({
       {Http::Headers::get().Method, "CONNECT"},
-      {Http::Headers::get().Host, hostname_},
+      {Http::Headers::get().Host, config_.hostname()},
   });
 
+  for (const auto& header : config_.headers()) {
+    if (header.append().value()) {
+      headers->addCopy(Http::LowerCaseString(header.header().key()), header.header().value());
+    } else {
+      headers->setCopy(Http::LowerCaseString(header.header().key()), header.header().value());
```

Comment:
do we think it should be legal to configure host header here?  Again I'm inclined to say no, but if we allow we should call it out explicitly IMO.

Reply:
Added comment to the proto to call it out.